src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -460,6 +460,20 @@ int sd_dhcp_client_set_mtu(sd_dhcp_client *client, uint32_t mtu) {
         return 0;
 }
 
+int sd_dhcp_client_set_anonymize(sd_dhcp_client *client, int anonymize) {
+        assert_return(client, -EINVAL);
+
+        client->anonymize = !!anonymize;
+
+        client->req_opts_size = ELEMENTSOF(default_req_opts_anonymize);
```

Comment:
So this will set `req_opts` to `default_req_opts_anonymize`, no matter what the value of `anonymize` is. This doesn't seem right.

But there's also a bigger problem. This call will reset all options, removing whatever was added using `sd_dhcp_client_set_request_option` previously. This is not right. Please preserve existing options. We know that if `client->anonymize` was false, the first block of options is `default_req_opts`, and if false, the first block of options is `default_req_opts_anonymize`. You should be able to make use of this to distinguish which options were set automatically.

Reply:
>> So this will set req_opts to default_req_opts_anonymize, no matter what the value of anonymize is. This doesn't seem right.

This function is called when   `dhcp_anonymize` is true. But can be assigned directly.

>But there's also a bigger problem. This call will reset all options, removing whatever was added using sd_dhcp_client_set_request_option previously.
 This is not right. Please preserve existing options. 

This just called after  'sd_dhcp_client_new' so defaults options are reset.

>We know that if client->anonymize was false, the first block of options is default_req_opts, and if false, the first block of options is default_req_opts_anonymize. You should be able to make use of this to distinguish which options were set automatically.

I can't follow you here .  If the `client->anonymize` is true then it resets the 'default_req_opts'  to `default_req_opts_anonymize`

Reply:
In existing usage, this *is* called right after `sd_dhcp_client_new`, but it is possible that in the future it will be called differently. Those functions are called `sd_dhcp_*` because they are intended to be exported in the future (and in fact are already used by external projects informally), so they must be written in a generic way.

> I can't follow you here . If the client->anonymize is true then it resets the 'default_req_opts' to default_req_opts_anonymize

If `sd_dhcp_client_set_anonymize(client, false)` is called, it'll still use `default_req_opts_anonymize`, which is wrong.

Reply:
>in existing usage, this is called right after sd_dhcp_client_new, but it is possible that in the future it will be called differently. 
Like how ?

Those functions are called sd_dhcp_* because they are intended to be exported in the future (and in fact are already used by external projects informally), so they must be written in a generic way.
I am aware of this fact.

>If sd_dhcp_client_set_anonymize(client, false) is called, it'll still use default_req_opts_anonymize, which is wrong.

you mean to say with the current API?